src/placement/mapless_jump_map.c

```diff
@@ -0,0 +1,1376 @@
+ *				information on whether or not an internal node
+ *				(non-target) in a domain has been used.
+ * \param[in]	layout		This is the current layout for the object.
+ *				This is needed for guaranteeing that we don't
+ *				reuse a target already in the layout.
+ * \param[in]	md		Object metadata used used to compare object
+ *				version with fail sequence.
+ *
+ * \return			Returns an error code if an error occurred,
+ *				otherwise 0.
+ */
+static int
+get_rebuild_target(struct pool_map *pmap, struct pool_target **target,
+		   uint64_t key, uint8_t *dom_used,
+		   struct pl_obj_layout *layout, struct daos_obj_md *md)
+{
+	uint8_t                 *used_tgts = NULL;
+	uint32_t                selected_dom;
+	uint32_t                fail_num = 0xFFc5;
+	uint32_t                try = 0;
+	struct pool_domain      *target_selection;
+	struct pool_domain      *root;
+	int rc = 0;
+
+	rc = pool_map_find_domain(pmap, PO_COMP_TP_ROOT, PO_COMP_ID_ALL, &root);
+	if (rc == 0) {
+		D_ERROR("Could not find root node in pool map.");
+		rc = -DER_NONEXIST;
+		return rc;
+	}
+
+	while (1) {
+
+		uint8_t skiped_targets = 0;
+		uint16_t num_bytes = 0;
+		uint32_t num_doms;
+
+		num_doms = root->do_child_nr;
+
+		uint64_t child_pos = (uint64_t)(root->do_children)
+				     - (uint64_t)root;
+		child_pos = child_pos / sizeof(struct pool_domain);
+
+		/*
+		 * If all of the nodes have been used for shards but we
+		 * still have shards to place mark all nodes as unused
+		 * so duplicates can be chosen
+		 */
+		if (isset_range(dom_used, child_pos, child_pos +
+				 num_doms - 1)) {
+			clrbit_range(dom_used, child_pos,
+				(child_pos + (num_doms - 1)));
+		}
+
+		/*
+		 * Choose domains using jump consistent hash until we find a
+		 * suitable domains that has not already been used.
+		 */
+		do {
+			key = crc(key, fail_num++);
+			selected_dom = jump_consistent_hash(key, num_doms);
+			target_selection = &(root->do_children[selected_dom]);
+		} while (isset(dom_used, (selected_dom + child_pos)));
+
+		/* Mark this domain as used */
+		setbit(dom_used, (selected_dom + child_pos));
+
+		/* To find rebuild target we examine all targets */
+		num_doms = target_selection->do_target_nr;
+
+		num_bytes = (num_doms / 8) + 1;
+
+		D_ALLOC_ARRAY(used_tgts, num_bytes);
+		if (used_tgts == NULL)
+			return -DER_NOMEM;
+
+		/*
+		 * Add the initial layouts targets to the bitmap of checked
+		 * targets.
+		 */
+		int i;
+		uint64_t start_pos;
+		uint64_t end_pos;
+
+		start_pos = (uint64_t)(&((target_selection->do_targets[0])));
+		end_pos = start_pos + (num_doms * sizeof(**target));
+
+		for (i = 0; i < layout->ol_nr; ++i) {
+			int id = layout->ol_shards[i].po_target;
+			uint64_t position;
+
+			pool_map_find_target(pmap, id, target);
+			position = (uint64_t)(*target);
+
+
+			if (position >= start_pos && position < end_pos) {
+				setbit(used_tgts, ((position - start_pos)
+				/ sizeof(**target)));
+				skiped_targets++;
+			}
+		}
```

Comment:
Not sure I get this part. So if there are targets already exists in this domain, then we should try next domain? or I miss sth?

Reply:
Yes.

If a given domain has been used before all other domains in that level have been used, we should use another domain.
For example if we have Racks [A B C] and have placed a target on rack A, we should skip rack A if we get that selection from JCH and instead try on rack B or C.

Reply:
Currently it does allow you to specify a minimum domain level to guarantee replication.

On map initialization you pass in a pl_map_init_attr that states what that domain level is and during placement if you don't have enough domains to satisfy that number it returns -DER_INVAL the same way the ring map does.

Otherwise it should just alliteratively place on the next available domain.